go/runtime/bundle/discovery.go

```diff
@@ -0,0 +1,452 @@
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"slices"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/oasisprotocol/oasis-core/go/common"
+	"github.com/oasisprotocol/oasis-core/go/common/crypto/hash"
+	"github.com/oasisprotocol/oasis-core/go/common/logging"
+	cmSync "github.com/oasisprotocol/oasis-core/go/common/sync"
+	"github.com/oasisprotocol/oasis-core/go/config"
+)
+
+// discoveryInterval is the time interval between (failed) bundle discoveries.
+const discoveryInterval = 15 * time.Minute
+
+// Discovery is responsible for discovering new bundles.
+type Discovery struct {
+	mu sync.RWMutex
+
+	startOne   cmSync.One
+	discoverCh chan struct{}
+
+	bundleDir      string
+	manifestHashes map[common.Namespace][]hash.Hash
+
+	globalBaseURLs  []string
+	runtimeBaseURLs map[common.Namespace][]string
+
+	registry Registry
+
+	logger logging.Logger
+}
+
+// NewDiscovery creates a new bundle discovery.
+func NewDiscovery(dataDir string, registry Registry) *Discovery {
+	logger := logging.GetLogger("runtime/bundle/discovery")
+
+	return &Discovery{
+		startOne:       cmSync.NewOne(),
+		bundleDir:      ExplodedPath(dataDir),
+		manifestHashes: make(map[common.Namespace][]hash.Hash),
+		discoverCh:     make(chan struct{}, 1),
+		registry:       registry,
+		logger:         *logger,
+	}
+}
+
+// Init sets up bundle discovery using node configuration and adds configured
+// and cached bundles to the registry.
+func (d *Discovery) Init() error {
+	// Consolidate all bundles in one place, which could be useful
+	// if we implement P2P sharing in the future.
+	if err := d.copyBundles(); err != nil {
+		return err
+	}
+
+	// Add copied and cached bundles to the registry.
+	if err := d.Discover(); err != nil {
+		return err
+	}
+
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	// Validate global repository URLs.
+	globalBaseURLs, err := validateAndNormalizeURLs(config.GlobalConfig.Runtime.Repositories)
+	if err != nil {
+		return err
+	}
+
+	// Validate each runtime's repository URLs.
+	runtimeBaseURLs := make(map[common.Namespace][]string)
+
+	for _, runtime := range config.GlobalConfig.Runtime.Runtimes {
+		urls, err := validateAndNormalizeURLs(runtime.Repositories)
+		if err != nil {
+			return err
+		}
+		if len(urls) == 0 {
+			continue
+		}
+		runtimeBaseURLs[runtime.ID] = urls
+	}
+
+	// Update discovery.
+	d.globalBaseURLs = globalBaseURLs
+	d.runtimeBaseURLs = runtimeBaseURLs
+
+	return nil
+}
+
+// Start starts the bundle discovery.
+func (d *Discovery) Start() {
+	d.startOne.TryStart(d.run)
+}
+
+// Stop halts the bundle discovery.
+func (d *Discovery) Stop() {
+	d.startOne.TryStop()
+}
+
+func (d *Discovery) run(ctx context.Context) {
+	d.logger.Info("starting discovery",
+		"dir", d.bundleDir,
+	)
+
+	ticker := time.NewTicker(discoveryInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+		case <-d.discoverCh:
+		case <-ctx.Done():
+			d.logger.Info("stopping discovery")
+			return
+		}
+
+		_ = d.Discover()
+		d.Download()
+	}
```

Comment:
In theory we could sent downloaded bundles (paths) over the channel directly. This way there would be no need to `os.ReadDir(d.bundleDir)`, with further check for every bundle if already part of registry,  for every tick? 

If we assume it is mandatory to call `init` before starting the `*Discovery` service, we could remove `d.Discover()` altogether from this function.

But then... the tick is every 10 minutes anyways so this is not really a performance bottleneck.

Reply:
Could probably do that, but not sure if it is worth. Will think about it.

What I want is to be able to call `Discover` and `Download` without running the discovery, which s useful for tests.

Reply:
Agree. Unless trivial to change feel free to keep as is :)